core/src/Kokkos_CheckUsage.hpp

```diff
@@ -0,0 +1,141 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_CHECK_USAGE_HPP
+#define KOKKOS_CHECK_USAGE_HPP
+
+#include <ostream>
+#include <sstream>
+#include <string>
+#include <type_traits>
+#include <variant>
+
+#include <Kokkos_Assert.hpp>
+#include <Kokkos_Macros.hpp>
+#include <Kokkos_ExecPolicy.hpp>
+
+using namespace std::literals;
+
+namespace Kokkos {
+
+[[nodiscard]] bool is_initialized() noexcept;
+[[nodiscard]] bool is_finalized() noexcept;
+
+namespace Impl {
+
+using ExecPolicyVariant =
+    std::variant<Kokkos::ParallelForTag, Kokkos::ParallelScanTag,
+                 Kokkos::ParallelReduceTag>;
+
+inline std::ostream& operator<<(std::ostream& os,
+                                ExecPolicyVariant exec_policy) {
+  std::visit(
+      [&os](auto&& arg) {
+        using ExecPolicyTagType = std::decay_t<decltype(arg)>;
+        if (std::is_same_v<Kokkos::ParallelForTag, ExecPolicyTagType>) {
+          os << "parallel_for";
+        } else if (std::is_same_v<Kokkos::ParallelScanTag, ExecPolicyTagType>) {
+          os << "parallel_scan";
+        } else if (std::is_same_v<Kokkos::ParallelReduceTag,
+                                  ExecPolicyTagType>) {
+          os << "parallel_reduce";
+        } else {
+          os << "";
+        }
+      },
+      exec_policy);
+
+  return os;
+}
+
+// This helper function generates a log string given a pack of
+// arguments that can be output to a stringstream
+template <typename... Ts>
+std::string generate_info(const Ts&... args) {
+  std::ostringstream ss;
+
+  constexpr auto num_args = sizeof...(Ts);
+
+  if constexpr (num_args != 0) {
+    ss << "Concerns";
+    ([&] { ss << " " << args; }(), ...);
+  }
+
+  return ss.str();
+}
+
+// The types CheckUsage and UsageRequires are for specifying run-time checks
+// of correct usage of Kokkos functions and types.
+template <typename... T>
+class CheckUsage;
+
+struct UsageRequires {
+  struct isInitialized {};
+  struct isNotFinalized {};
+  struct insideExecEnv {};
+};
+
+template <>
+class CheckUsage<UsageRequires::isInitialized> {
+ public:
+  static inline void check() {
+    KOKKOS_EXPECTS_CRITICAL(CHECK_CONDITION(Kokkos::is_initialized()));
```

Comment:
If `KOKKOS_EXPECTS_CRITICAL` is only used in these `check` member functions, there isn't much of a benefit for using a macro instead of a regular function.